src/primitives/compression.h

```diff
@@ -0,0 +1,475 @@
+        uint8_t lockTime = 0;
+        if (this->m_shortendLockTime) lockTime = 1; else if (this->m_nLockTime != 0) lockTime = 2;
+
+        uint64_t control = version;
+        control |= inputCount << 2;
+        control |= outputCount << 4;
+        control |= lockTime << 6;
+        s << VARINT(control);
+
+        if (!version) s << VARINT(this->m_nVersion);
+        if (!inputCount) s << VARINT(this->m_vin.size());
+        if (!outputCount) s << VARINT(this->m_vout.size());
+        if (lockTime) s << VARINT(this->m_nLockTime);
+
+        for (size_t index = 0; index < std::max(this->m_vin.size(), this->m_vout.size()); index++) {
+            uint64_t vControl = 0;
+            if (this->m_vin.size() > index) vControl = this->m_vin.at(index).GetMetadata();
+            uint8_t serializedScriptType = 0;
+            if (this->m_vout.size() > index) serializedScriptType = this->m_vout.at(index).GetSerializedScriptType();
+            vControl |= serializedScriptType << 8;
+            s << VARINT(vControl);
+            if (this->m_vin.size() > index) this->m_vin.at(index).Serialize(s);
+            if (this->m_vout.size() > index) this->m_vout.at(index).Serialize(s);
+        }
+    }
+
+    template <typename Stream>
+    inline void Unserialize(Stream& s) {
+        uint64_t control = std::numeric_limits<uint64_t>::max();
+        s >> VARINT(control);
+        uint8_t version = control & 0b11;
+        uint8_t inputCount = (control & (0b11 << 2)) >> 2;
+        uint8_t outputCount = (control & (0b11 << 4)) >> 4;
+        uint8_t lockTime = (control & (0b11 << 6)) >> 6;
+        uint32_t nInputCount = std::numeric_limits<uint32_t>::max();
+        uint32_t nOutputCount = std::numeric_limits<uint32_t>::max();
+
+        if (!version) {
+            this->m_nVersion = std::numeric_limits<uint32_t>::max();
+            s >> VARINT(this->m_nVersion);
+        } else this->m_nVersion = version;
+        if (!inputCount) {
+            s >> VARINT(nInputCount);
+        } else nInputCount = inputCount;
+        if (!outputCount) {
+            s >> VARINT(nOutputCount);
+        } else nOutputCount = outputCount;
+        this->m_shortendLockTime = false;
+        if (lockTime) {
+            if (lockTime ==    1) this->m_shortendLockTime = true;
+            this->m_nLockTime = std::numeric_limits<uint32_t>::max();
+            s >> VARINT(this->m_nLockTime);
+        } else this->m_nLockTime = lockTime;
+
+        for (size_t index = 0; index < std::max(nInputCount, nOutputCount); index++) {
+            uint64_t vControl = std::numeric_limits<uint64_t>::max();
+            s >> VARINT(vControl);
+            uint8_t serializedScriptType = (vControl & (0b111 << 8)) >> 8;
+
+            if (nInputCount > index) this->m_vin.push_back(CCompressedTxIn(deserialize, s, vControl));
+            if (nOutputCount > index) this->m_vout.push_back(CCompressedTxOut(deserialize, s, serializedScriptType));
```

Comment:
In 89b52782f8d6aca75d49048da1e137fbf615c0e6:

I'm curious why you decided to interleave inputs and outputs. I think it would be clearer to have them one after the other.

Reply:
It saves space on the compression byte that precedes the input output pairs.

Reply:
I only need 2-3 bits for the outputs in the preceding byte if I can group it with the input metadata I can save 6-7 bits on average.

Reply:
RIght -- it's fine to group the metadata together, but I'd still serialize the inputs themselves all together, followed by the outputs all together.

Reply:
Fixed, Thanks!